(GH-2323)(Docs) Finalize module docs

hestonhoffman opened the pull request:
Closes #2323

Major changes:
- Restructured the modules section to emphasize "Modules overview" to
  readers. These parent pages can get a bit lost in the nav.
- Turned `modules.md` into "Modules overview" and added some high-level
  explanation of modules (in Bolt) and dependency management.
- Edits to Module structure doc. This doc lacked direction. I've tried
  to narrow the focus down to modules that you're using with Bolt -
  where you should put the code and how you can load it with Bolt.
- Removed the Managing modules page and added most of the content from
  that page to "Installing modules".
- Moved the "What's changed" section to the relevant developer update.
- Moved the packaged modules section to "Modules overview."
- Moved the compatibility stuff to "Modules overview."
- Removed/Deprecated most references to the old Puppetfile workflows.
- Moved the "Migrate a Bolt project" info to the projects page.
- Added a section on modules to Troubleshooting.

For an exhaustive list of changes, see #2323

!no-release-note

beechtom commented on documentation/projects.md:
````diff
@@ -271,10 +273,127 @@ BOLT_PROJECT='~/project/my_project' bolt command run uptime -t target1
+      - uri: target2.example.com
+        alias: target2
+```
+
````
```suggestion

### How Bolt updates project configuration

Bolt locates a `bolt.yaml` file and moves its configuration to a new `bolt-project.yaml` file
and the `inventory.yaml` file. It then deletes the `bolt.yaml` file.

Project-specific configuration is moved to the `bolt-project.yaml` file, while transport
configuration is moved to the top-level `config` key in the `inventory.yaml` file. If the
`inventory.yaml` file has an existing top-level `config` key, the transport configuration
from `bolt.yaml` is deep merged, with the configuration in `inventory.yaml` having higher
precedence.

For example, a project with the following `bolt.yaml` and `inventory.yaml` files:

'''yaml
# bolt.yaml
format: json
transport: winrm
winrm:
  user: Administrator
  password: Bolt!
'''

'''yaml
# inventory.yaml
groups:
  - name: windows
    targets:
      - target1.example.com
      - target2.example.com
config:
  winrm:
    ssl: false
'''

Becomes:

'''yaml
# bolt-project.yaml
format: json
'''

'''yaml
# inventory.yaml
groups:
  - name: windows
    targets:
      - target1.example.com
      - target2.example.com
config:
  transport: winrm
  winrm:
    ssl: false
    user: Administrator
    password: Bolt!
'''

```

lucywyman commented on documentation/bolt_installing_modules.md:
```diff
@@ -50,6 +54,284 @@ mod 'puppetlabs-puppetserver_gem', '1.1.1'
+
+| Key | Description | Required |
+| --- | --- | :-: |
+| `name` | The full name of the module. Can be either `<OWNER>-<NAME>` or `<OWNER>/<NAME>`. | ✓ |
```
!!! The check mark! I love it.

hestonhoffman replied:
I think Tom did that 😄

lucywyman commented on documentation/bolt_running_tasks.md:
```diff
@@ -102,19 +102,17 @@ You can also load parameters from a file by putting `@` before the file name.
+modulepath. Bolt loads the current [Bolt
+project](./experimental_features.md#bolt-projects) as a standalone module at the
+front of the modulepath. If you are developing a new task, you can create a
+`<PROJECT DIRECTORY>/bolt-project.yaml` file, develop your task in `<PROJECT
```
Should out `bolt project init` while we're here?

hestonhoffman replied:
Also simplifying this language a bit

lucywyman commented on documentation/developer_updates.md:
```diff
@@ -36,6 +36,30 @@ install a Puppetfile, because Bolt installs modules to a special directory that
+project](bolt_installing_modules.md#create-a-bolt-project-with-pre-installed-modules)
+or [migrate an existing project](./projects.md#migrate-a-bolt-project).
+
+Here's a summary of what's changed:
```
I think it'd be best to point out that `bolt project migrate` will do a lot of this for you towards the top here, but I'm not sure where to put it. Basically I think it's "Run `bolt project migrate` to use the new workflow. You shouldn't have to think about it, but here's what changed if you're interested"?

lucywyman commented on documentation/migrating_inventory_files.md:
````diff
@@ -78,4 +78,4 @@ groups:
       - uri: target2.example.com
         alias: target2
-```
+```
````
Lol Git.

hestonhoffman replied:
IKR?

hestonhoffman commented on documentation/modules.md:
```diff
@@ -1,16 +1,220 @@
+You can learn more about the module changes introduced in Bolt 2.23.0 in the
+[September 2020 developer update](./developer_updates.md#september-2020).
+
+## Packaged modules
```
Definitely, I'll add that!

lucywyman replied:
I actually think pointing people to the Puppetfile might be the best way to do this, rather than manually listing out all the modules? Mostly because I think we'd definitely forget to update this if we added new modules.

hestonhoffman replied:
Hmmm. I actually like the idea of replacing the list of modules with a link to a generated file. In that case I could link to both a supported plugins page and a supported modules page. Let's ticket that and do it later, though. For now I'll just find a place to link to supported plugins.

hestonhoffman replied:
https://github.com/puppetlabs/bolt/issues/2382

lucywyman commented on documentation/modules.md:
```diff
@@ -1,16 +1,219 @@
+├── resource_api/
+├── stdlib/
+└── translate/
+
```
```suggestion
```
```
This actually seems to render fine in Dita, but makes the whole next paragraph a code block in markdown

lucywyman commented on documentation/bolt_installing_modules.md:
```diff
@@ -50,6 +54,284 @@ mod 'puppetlabs-puppetserver_gem', '1.1.1'
+
+Follow these steps to pin a module version:
+1. Find the module under the `modules` key in your `bolt-project.yaml` file.
+2. Add a `version_requirement` key that specifies the version requirements for
```
```suggestion
1. Add a `version_requirement` key that specifies the version requirements for
```

lucywyman commented on documentation/bolt_installing_modules.md:
````diff
@@ -50,6 +54,284 @@ mod 'puppetlabs-puppetserver_gem', '1.1.1'
+   - name: puppetlabs/apache
+     version_requirement: '>= 4.0.0 < 6.0.0'
+   ```
+3. Run the following command. The `force` option is required because you've made
````
```suggestion
1. Run the following command. The `force` option is required because you've made
```

hestonhoffman commented on documentation/bolt_installing_modules.md:
```diff
@@ -50,6 +54,284 @@ mod 'puppetlabs-puppetserver_gem', '1.1.1'
+- [bolt-project.yaml options](bolt_project_reference.md#module-install)
+- [bolt-defaults.yaml options](bolt_defaults_reference.md#module-install)
+
+## ⛔ Manage modules with a Puppetfile
+
+⛔ **DEPRECATED:** This method of installing and managing modules is deprecated.
+Use the workflows outlined above and avoid editing your Puppetfile.
```
Ah interesting! I guess we can just document that stuff as it comes in then

hestonhoffman commented on documentation/managing_modules.md:
```diff
@@ -1,398 +0,0 @@
-# 🧪 Managing modules in Bolt projects
```
Yes, I've got a PR up in the web repo. I'm just redirecting it to the installing page. Maybe we can update the short link to point to that page?

hestonhoffman commented on documentation/module_structure.md:
````diff
@@ -1,25 +1,20 @@
+documentation](https://puppet.com/docs/puppet/latest/modules_fundamentals.html#module_structure).
+However, a typical module for use with Bolt may contain these directories:
 
 ```console
 ├── data/
````
Actually, I think I'm just going to remove this tree. It's redundant with the table anyway

hestonhoffman commented on documentation/projects.md:
```diff
@@ -271,10 +273,186 @@ BOLT_PROJECT='~/project/my_project' bolt command run uptime -t target1
+
+For example, a project with the following `bolt.yaml` and `inventory.yaml` files:
+
+'''yaml
```
Euugh. How did that happen? A bunch of these are single quotes instead of backticks. Fixed!

beechtom commented on documentation/bolt_installing_modules.md:
```diff
@@ -1,28 +1,32 @@
+with the `puppetlabs/apache` and `puppetlabs/mysql` modules installed, use the following command:
 
-- _\*nix shell command_
+_\*nix shell command_
```
I can't add a suggestion, unfortunately, but the PowerShell cmdlet a couple lines down should also be un-bulleted.

beechtom commented on documentation/bolt_installing_modules.md:
```diff
@@ -50,6 +54,284 @@ mod 'puppetlabs-puppetserver_gem', '1.1.1'
+1. Find the module under the `modules` key in your `bolt-project.yaml` file.
+1. Add a `version_requirement` key that specifies the version requirements for
+   the module. For example, the following `bolt-project.yaml` file sets the
+   version requirement for the `apache` module to greater than or equal to
```
```suggestion
   version requirement for the `puppetlabs/apache` module to greater than or equal to
```

beechtom commented on documentation/developer_updates.md:
````diff
@@ -56,6 +56,24 @@ management! With this improvement, you no longer need to manually manage your
+Update-BoltProject
+```
+
+### Why are we making these changes?
````
```suggestion
#### Why are we making these changes?
```

Otherwise this will show in the ToC as its own update.

beechtom commented on documentation/projects.md:
```diff
@@ -140,20 +144,23 @@ The following are common files and directories found in a Bolt project.
-|[`Puppetfile`](bolt_installing_modules.md#)|Specifies which modules to install for the project. 🧪 **Experimental:** If you've opted in to the experimental module management workflows, Bolt manages this file. Avoid editing it. For more information, see [Managing modules](managing_modules.md).|
-|[`modules/`](bolt_installing_modules.md#)|The directory where modules from the `Puppetfile` are installed. In most cases, do not edit these modules locally.|
-|[`site-modules/`](bolt_installing_modules.md)|Local modules that are edited and versioned with the Bolt directory.|
+|`files/`|A directory for storing content consumed by your tasks and plans, such as scripts.|
```
Can't make a suggestion since it's too far up, but the 'Directory' column should be renamed to 'Directory/File' as well.

hestonhoffman commented:
Okay so I removed the alt forge and proxy bits, and added the project init command to the GS. The GS needs more updates, but I have a separate ticket for that on the Kanban, so I'll get back to it after Thanksgiving :)

beechtom reviewed (approved):
This is great! 🙌
